File: 01_process_thread/python/main.py
import os
import time
from datetime import datetime, timezone

from fastapi import FastAPI, Query as QueryParam
from fastapi.responses import JSONResponse

app = FastAPI()

HEAVY_CALC_N = int(os.environ.get("HEAVY_CALC_N", "38"))


def fibonacci(n: int) -> int:
    if n <= 1:
        return n
    return fibonacci(n - 1) + fibonacci(n - 2)


def utc_now_iso() -> str:
    now = datetime.now(timezone.utc)
    return now.strftime("%Y-%m-%dT%H:%M:%S.") + f"{now.microsecond // 1000:03d}Z"


@app.get("/health")
async def health() -> JSONResponse:
    return JSONResponse({"status": "ok", "language": "python"})


@app.get("/heavy")
async def heavy(n: int | None = QueryParam(default=None)) -> JSONResponse:
    calc_n = n if n is not None else HEAVY_CALC_N
    started_at = utc_now_iso()
    start = time.monotonic()

    fibonacci(calc_n)

    duration_ms = round((time.monotonic() - start) * 1000)
    finished_at = utc_now_iso()

    return JSONResponse({
        "language": "python",
        "threadId": f"uvicorn-worker-{os.getpid()}",
        "startedAt": started_at,
        "finishedAt": finished_at,
        "durationMs": duration_ms,
    })

File: 01_process_thread/go/main.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"runtime"
	"strconv"
	"strings"
	"time"
)

func getGoroutineID() string {
	var buf [64]byte
	n := runtime.Stack(buf[:], false)
	s := string(buf[:n])
	// format: "goroutine 123 [running]:\n..."
	s = strings.TrimPrefix(s, "goroutine ")
	if idx := strings.IndexByte(s, ' '); idx > 0 {
		return s[:idx]
	}
	return "unknown"
}

func fibonacci(n int) int {
	if n <= 1 {
		return n
	}
	return fibonacci(n-1) + fibonacci(n-2)
}

func main() {
	heavyCalcN := 40
	if v := os.Getenv("HEAVY_CALC_N"); v != "" {
		if parsed, err := strconv.Atoi(v); err == nil {
			heavyCalcN = parsed
		}
	}

	mux := http.NewServeMux()

	mux.HandleFunc("GET /health", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]string{
			"status":   "ok",
			"language": "go",
		})
	})

	mux.HandleFunc("GET /heavy", func(w http.ResponseWriter, r *http.Request) {
		n := heavyCalcN
		if qn := r.URL.Query().Get("n"); qn != "" {
			if parsed, err := strconv.Atoi(qn); err == nil {
				n = parsed
			}
		}

		startedAt := time.Now()

		fibonacci(n)

		finishedAt := time.Now()
		durationMs := finishedAt.Sub(startedAt).Milliseconds()

		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]interface{}{
			"language":   "go",
			"threadId":   fmt.Sprintf("goroutine-%s", getGoroutineID()),
			"startedAt":  startedAt.UTC().Format(time.RFC3339Nano),
			"finishedAt": finishedAt.UTC().Format(time.RFC3339Nano),
			"durationMs": durationMs,
		})
	})

	fmt.Printf("Go server listening on :8080 (GOMAXPROCS=%d)\n", runtime.GOMAXPROCS(0))
	if err := http.ListenAndServe(":8080", mux); err != nil {
		fmt.Fprintf(os.Stderr, "server error: %v\n", err)
		os.Exit(1)
	}
}
